### PR TITLE
Isolate Role Authorization

### PR DESCRIPTION
**Pull request by mccallofthewild**
Currently role authorization is redundant and stored locally across every contract: 

image

This PR isolates role storage to a single contract and implements a Controller/Provider pattern for easy use and isolated testing.

### CODE REVIEW

File: contracts/active-pool/src/state.rs [hunk 1]
```diff
@@ -23,6 +16,17 @@ pub struct SudoParams {
     pub owner: Addr,
 }
 
+pub struct State<'a> {
```
Comment (vernonjohnson) [inline contracts/active-pool/src/state.rs:19]: Hmm... I'm a bit confused about the usage of State. It needs the role provider address to instantiate state.roles (which has type RoleConsumer). But everywhere it's used, you're calling State::default which passes in a dummy "role_provider_address".

Shouldn't we be doing the following instead:
1) Store State as an item
```
pub const STATE: Item<State> = Item::new("state");
 ```
2. Creating new State object with the actual role_provider address in contract instantiation
```
   let role_provider =  deps.api.addr_validate(&msg.role_provider)?;
   let state = State {
        roles: RoleConsumer::new(role_provider),
    };
    STATE.save(deps.storage, &state)?;
```
TBH I'm a bit confused in general about the RoleConsumer usage. Wondering if there's a simpler way to accomplish the same thing

Comment (mccallofthewild) [reply contracts/active-pool/src/state.rs:19]: @vernonjohnson `State` is a struct of storage objects that manages contract state. This pattern makes it easier to implement lifetimes for storage types.

`RoleConsumer` is struct that wraps `Item::new("role_provider_address")` and provides common functionality to query the role provider contract via its given address.

File: packages/ultra-base/src/active_pool.rs [hunk 2]
```diff
@@ -11,22 +11,9 @@ pub struct InstantiateMsg {
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
```
[the remark this answers is not in the document]
Comment (mccallofthewild) [reply packages/ultra-base/src/active_pool.rs:11]: I agree. Let's tackle this next. Opened an issue https://github.com/notional-labs/UltraStableJuno/issues/25

File: packages/ultra-controllers/src/roles.rs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    /// Returns Ok(true) if this user has any of the roles, Ok(false) if not and an Error if
+    /// we hit an error with Api or Storage usage
+    pub fn has_any_role(&self, store: &dyn Storage, roles: &[Role], caller: &Addr) -> StdResult<bool> {
+        for role in roles {
+           if self.has_role(store, role, caller)? {
+            // if any exists, stop iteration and return true result
+             return Ok(true)
+           } else {
```
[the remark this answers is not in the document]
Comment (mccallofthewild) [reply packages/ultra-controllers/src/roles.rs]: Fixed in https://github.com/notional-labs/UltraStableJuno/pull/24/commits/8da9b7fd13556322919471e976100769c1daff06 👍

File: contracts/ultra-role-provider/.circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
+version: 2.1
```
[the remark this answers is not in the document]
Comment (mccallofthewild) [reply contracts/ultra-role-provider/.circleci/config.yml:1]: The DAODAO workflows would be good. This is all boilerplate from the Interwasm cw template

**Comment by mccallofthewild** [comment]
Triple checked that all role requirements were migrated accurately. Merging.

**Pull request merged**